proof-systems/src/darlin/pcd/mod.rs

```diff
@@ -0,0 +1,232 @@
+    pub segment_size: usize
+}
+
+impl PCDParameters {
+    /// We assume the DLOG keys to be generated outside the PCD scheme,
+    /// so this function actually just trim them to the segment size
+    /// specified in the config.
+    pub fn universal_setup<G: AffineCurve, D: Digest>(
+        &self,
+        params: &UniversalParams<G>
+    ) -> Result<(DLogCommitterKey<G>, DLogVerifierKey<G>), PCError>
+    {
+        InnerProductArgPC::<G, D>::trim(
+            params,
+            self.segment_size - 1,
+        )
+    }
+}
+
+/// Trait for the recursive circuit of a PCD node in G. Both witnesses and public inputs
+/// are derived from previous proofs (PCDs) and some additional incremental data ("payload").
+/// A recursive circuit comes with a universal circuit interface, comprised of 
+///     - `user inputs` (i.e. the proof "statement") and 
+///     - `system inputs`, which is the data due to amortization and split verification, 
+///     aka deferred checks. 
+/// The incremental data is used only by dedicated circuits such as a base proofs or
+/// a finalizing block proofs. For the ordinary merger nodes, it is simply `None`.
+pub trait PCDCircuit<G: AffineCurve>: ConstraintSynthesizer<G::ScalarField> {
+
+    /// Any data that may be needed to bootstrap the circuit that is not covered by the other
+    /// fields. 
+    type SetupData: Clone;
+
+    /// Additional data to be processed by the circuit. 
+    /// This might be related to recursion (incremental "payload"). In our PCD it is  
+    /// supplementary witness data to serve additional business logic of the circuit.
+    type IncrementalData;
+
+    /// Elements that are deferred during recursion. The are derived from the PCDs 
+    /// passed by the nodes "below" 
+    type SystemInputs: ToConstraintField<G::ScalarField> + Debug + Clone;
+
+    /// PCD type the circuit needs to verify
+    type PreviousPCD:  PCD;
+
+    /// Initialize the circuit state without explicitly assigning inputs and witnesses.
+    /// To be used to generate pk and vk.
+    fn init(config: Self::SetupData) -> Self;
+
+    /// Assign a concrete state to the circuit, using previous proofs and some "payload".
+    /// As the circuit needs to verify previous proofs, it also needs the corresponding vks;
+    fn init_state(
+        config:               Self::SetupData,
+        previous_proofs_data: Vec<Self::PreviousPCD>,
+        previous_proofs_vks:  Vec<<Self::PreviousPCD as PCD>::PCDVerifierKey>,
+        incremental_data:     Self::IncrementalData,
+    ) -> Self;
+
+    /// Extract the system inputs from a concrete instantiation of the circuit.
+    /// Return Error if it's not possible to derive SystemInputs.
+    fn get_sys_ins(&self) -> Result<&Self::SystemInputs, PCDError>;
+
+    /// Extract the user inputs from a concrete instantiation of the circuit.
+    /// Return Error if it's not possible to derive UserInputs.
+    fn get_usr_ins(&self) -> Result<Vec<G::ScalarField>, PCDError>;
```

Comment:
It might be useful to have a separate ``fn get_circuit_inputs()``, as appending ``sys_ins`` and ``usr_ins`` (a now done in the FinalDarlin prover) might be not useful/flexible enough.

Reply:
Put a TODO in the code